### PR TITLE
Adding Windows Overlay support to Kube Proxy

### PR DESCRIPTION
**Pull request by ksubrmnn**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
 /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR adds support in Kube-proxy for Overlay networks in Windows. Kube-Proxy currently only supports L2Bridge networks on Windows. On Windows, Host Networking Service (HNS) is responsible for ensuring network connectivity. HNS needs to be programmed in a different way for Overlay than L2Bridge.

Specifically, HNS needs additional configuration to allow Cluster IP connectivity. This is reflected in this PR by programming "service VIPs."

Currently HNS supports both DSR and non DSR for loadbalancing. This support is also reflected in this PR by adding support for both.

This PR also adds support for using HNS in V1 and V2. V2 is necessary for DSR and Overlay, but non DSR L2Bridge is supported in both.

Testing:
Basic unit tests for Kube Proxy are added in this PR to check if remote endpoints and loadbalancers are created and accounted for. More testing will follow in subsequent PRs.

Tests were also created to test HNS V1 and V2 implementations that were added in this PR. 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This is an Alpha feature
There is no design doc associated with this feature
Hcsshim was updated to include the RemoteSubnetPolicy API changes

**Does this PR introduce a user-facing change?**:

```release-note
Added kube-proxy support for overlay networking and DSR in Windows and new flags for network-name, source-vip, and enable-dsr.
```

### CODE REVIEW

**Comment by thockin** [comment]
Can you please update the first comment in this PR to include more information?  A lot more.

What does this do?  Pretend I don't know anything about Windows.

Was there a design doc or KEP?

Is it considered Alpha or Beta or GA?

Where are the tests?  That's a LOT of code to add with no tests.

What should the release notes say about this?

I can't do meaningful review of it because I really don't know what it is doing.

**Comment by madhanrm** [comment]
@ksubrmnn  can you resolve the merge conflict?

File: cmd/kube-proxy/app/server.go [hunk 1]
```diff
@@ -176,6 +176,9 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
 		"A string slice of values which specify the addresses to use for NodePorts. Values may be valid IP blocks (e.g. 1.2.3.0/24, 1.2.3.4/32). The default empty string slice ([]) means to use all local addresses.")
 	fs.Var(flag.NewMapStringBool(&o.config.FeatureGates), "feature-gates", "A set of key=value pairs that describe feature gates for alpha/experimental features. "+
 		"Options are:\n"+strings.Join(utilfeature.DefaultFeatureGate.KnownFeatures(), "\n"))
+	fs.StringVar(&o.config.Winkernel.SourceVip, "source-vip", o.config.Winkernel.SourceVip, "The IP address of the source VIP for non-DSR.")
```
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: We need to move OS-specific flags to an OS-specific place so they don't appear on other OSes.

Comment (thockin) [reply cmd/kube-proxy/app/server.go]: in fact, we seem to have this already with `addOSFlags`

**Comment by madhanrm** [comment]
Overall comment is to split WinKernel proxy code to have 2 interfaces for L2Bridge & Overlay, to avoid lot of If/else condition.
/lgtm

File: pkg/proxy/winkernel/proxier.go [hunk 2]
```diff
@@ -479,19 +505,73 @@ func NewProxier(
...
 	klog.V(1).Infof("Hns Network loaded with info = %v", hnsNetwork)
+	var isDSR bool
+	isDSR = config.EnableDSR
+
+	var sourceVip string
+	var hostMac string
+	if hnsNetwork.Type == "Overlay" {
+		supportedFeatures := hcn.GetSupportedFeatures() //TODO Create RemoteSubnet Feature
+		if supportedFeatures.Api.V2 {
+			hns = hnsV2{}
```
Comment (fasaxc) [inline pkg/proxy/winkernel/proxier.go]: HNS v2 is required for the DSR feature, which is required (by Calico) with the l2bridge network type.  Can this check be moved out of the Type == Overlay conditional?

Comment (fasaxc) [reply pkg/proxy/winkernel/proxier.go]: Looks like the blocking PR has now been merged? https://github.com/Microsoft/hcsshim/pull/436

**Comment by ksubrmnn** [comment]
@thockin Can you take a look at the first 3 commits please? The rest are Windows related, but the first 3 pertain to vendoring and config changes.

**Comment by thockin** [comment]
If it is alpha it should be covered by a feature gate, shouldn't it?

Otherwise looks OK to me (well contained anyway).  Will wait for LGTM from @madhanrm and proper relnote.  Ping me when ready for final.

File: pkg/proxy/winkernel/proxier.go [hunk 3]
```diff
@@ -988,44 +1006,75 @@ func (proxier *Proxier) syncProxyRules() {
...
-				}
-
-				newHnsEndpoint, err = hnsnetwork.CreateRemoteEndpoint(hnsEndpoint)
-				if err != nil {
-					klog.Errorf("Remote endpoint creation failed: %v", err)
-					continue
+				if proxier.network.networkType == "Overlay" {
+					networkV2, err := hns.getNetworkByName(hnsNetworkName)
+					if err != nil {
+						klog.Errorf("%v", err)
```
[the remark this answers is not in the document]
Comment (ksubrmnn) [reply pkg/proxy/winkernel/proxier.go]: Will revamp error handling in follow up PR

File: pkg/proxy/winkernel/proxier.go [hunk 4]
```diff
@@ -988,44 +1006,75 @@ func (proxier *Proxier) syncProxyRules() {
...
+					}
+
+					var providerAddress string
+					for _, rs := range networkV2.remoteSubnets {
+						_, ipNet, err := net.ParseCIDR(rs.destinationPrefix)
+						if err != nil {
+							klog.Errorf("%v", err)
```
[the remark this answers is not in the document]
Comment (ksubrmnn) [reply pkg/proxy/winkernel/proxier.go]: Will revamp error handling in follow up PR

File: pkg/proxy/winkernel/hnsV2.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+		return nil, err
+	}
+
+	var remoteSubnets []*remoteSubnetInfo
+	for _, policy := range hnsnetwork.Policies {
+		if policy.Type == hcn.RemoteSubnetRoute {
+			policySettings := hcn.RemoteSubnetRoutePolicySetting{}
+			err = json.Unmarshal(policy.Settings, &policySettings)
+			if err != nil {
+				klog.Infof("Failed to unmarshal settings")
```
[the remark this answers is not in the document]
Comment (ksubrmnn) [reply pkg/proxy/winkernel/hnsV2.go]: Will revamp error handling in follow up PR

**Comment by ksubrmnn** [comment]
@thockin Do we still need a feature flag if we are checking to see if the network they are providing Kube Proxy is overlay or not? In Windows, the a network needs to be created before kube-proxy is started. This network's type is checked, and based on that, we program for either l2bridge of overlay. Since we already have this check, maybe a feature flag is not needed? I included a commit that adds it anyway, but let me know if you think it's still necessary.

**Comment by madhanrm** [comment]
ping @thockin

**Comment by thockin** [comment]
If this is considered "alpha" it needs a feature gate.  Unless that gate is enabled, it should behave as if this feature is not present.  We want to avoid users getting a surprise.  I understand that they would have to have configured for this explicitly, so it should be OK to also enable the gate, right?

**Comment by ksubrmnn** [comment]
@thockin Added the feature gates for WinOverlay and WinDSR. All the checks have passed when I open the PR dashboard, not sure why it's not updated in GitHub, but this is PR is ready.

**Comment by madhanrm** [comment]
@thockin  can you please take a look at this PR and help us get this merged? We have server follow up PRs on top of this one.

**Comment by dineshgovindasamy** [comment]
@thockin following up from the sig/network meeting, Could you please take a look at this PR and provide your feedback or approve the PR?

**Comment by ksubrmnn** [comment]
@thockin I've separated out the Windows only code into commit b724bdb19a8a16ca6ae628f9bee2c49ee9084f7b.

You only need to look at the first 3 commits and the last one. I believe you've already looked at the first 3 so only the last one is new and includes the feature flags.

**Pull request merged**